### PR TITLE
GPU implementation of the Mean Squared Error loss function

### PR DESCRIPTION
**Pull request by ghost**
This PR, along with the [one on the _dll_ repository](https://github.com/wichtounet/dll/pull/18) adds support for the Mean Squared Error loss function on GPU.

I used the BCE files as base, as I thought the structure could be carried over.

### CODE REVIEW

File: src/mse.cu [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+
+            if (i + blockSize < n) {
+                mySum += mse_error(output[(i + blockSize) * incx], labels[(i + blockSize) * incx]);
+            }
+
+            i += gridSize;
+        }
+    }
+
+    shared_data[tid] = mySum / n;
```
Comment (wichtounet) [inline src/mse.cu]: Why did you divide by `n` here? By doing this, the implementation will not be equivalent to the one from DLL. The scaling is done by using the `alpha `and `beta `factors.

File: src/mse.cu [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        T* host_output = new T[n];
+        T* host_labels = new T[n];
+
+        cuda_check(cudaMemcpy(host_output, output, n * sizeof(T), cudaMemcpyDeviceToHost));
+        cuda_check(cudaMemcpy(host_labels, labels, n * sizeof(T), cudaMemcpyDeviceToHost));
+
+        for (size_t i = 0; i < n; i++) {
+            result += (host_labels[i] - host_output[i]) * (host_labels[i] - host_output[i]);
+        }
+        result /= n;
```
Comment (wichtounet) [inline src/mse.cu]: Same here.

File: src/mse.cu [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+
+        delete[] host_output;
+        delete[] host_labels;
```
Comment (wichtounet) [inline src/mse.cu:413]: There should be a `return result` here. If you want, you can also fix it in BCE where it's missing.

File: src/mse.cu [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+
+        cuda_check(cudaMemcpy(host_output, output, n * sizeof(T), cudaMemcpyDeviceToHost));
+        cuda_check(cudaMemcpy(host_labels, labels, n * sizeof(T), cudaMemcpyDeviceToHost));
+
+        for (size_t i = 0; i < n; i++) {
+            loss += (host_labels[i] - host_output[i]) * (host_labels[i] - host_output[i]);
+            error += fabsf(host_labels[i] - host_output[i]);
+        }
+        loss /= n;
+        error /= n;
```
Comment (wichtounet) [inline src/mse.cu]: Same here

**Review by wichtounet — COMMENTED** [review]
Seems good at first sight 👍 

Just a few comments to change. 

Can you comment on the speedups compared without it? 

I think your forgot to submit the PR for `etl`.

**Comment by ghost** [comment]
These are the performance results (with `no_bias_display` and `no_epoch_error` disabled to compute the losses and errors):

MSE bench [ms] | std | std | std | GPU | GPU | GPU | % improvement (best times)
-- | -- | -- | -- | -- | -- | -- | --
mnist_ae_perf | 8248 | 7575 | 7492 | 6772 | 5913 | 5937 | 21.0758142
mnist_deep_ae_perf | 9869 | 9832 | 9769 | 10289 | 9479 | 9494 | 2.968574061
mnist_cnn_perf | 10879 | 10168 | 10180 | 10963 | 10108 | 10106 | 0.6097560976
mnist_dbn_perf | 16838 | 16181 | 16257 | 16078 | 16024 | 16082 | 0.9702737779
mnist_mlp_perf | 8830 | 8095 | 8082 | 8747 | 7947 | 7968 | 1.670378619
cifar10_cnn_small | 12151 | 11374 | 11410 | 12123 | 11413 | 11432 | -0.3428872868
cifar10_cnn_med | 19706 | 19654 | 19601 | 19444 | 19531 | 19580 | 0.8009795419
cifar10_cnn_big | 37742 | 37774 | 37936 | 37815 | 38134 | 37663 | 0.2093158815

I expected better results. The only network that gained is the Autoencoder (and to a lesser extent the Deep Autoencoder), with still just a 20% improvement.
The DLL timings show that the MSE takes up a lot of time still. I will fix the issues discussed and retest, maybe that will change something.

**Comment by ghost** [comment]
One thing I noticed is that, even with the `no_batch_display` and `no_epoch_error` parameters deactivated, the ETL counters don't show any D <-> H memory copy anymore, as it used to happen before.

If we print data, shouldn't it be copied over to the CPU and therefore be visible in the counters?

This is why I asked you if you had made any changes to ETL, specifically to the Expressions, that allow the expression to be computed more efficiently on GPU?

Example counters for the `dll_mnist_ae_perf` network, with `dll_mean_squared_error` and `define ETL_GPU_POOL`:
 ```
---------------------------
 | Counter        | Count  |
 ---------------------------
 | egblas         | 162050 |
 | gpu:assign     | 64562  |
 | impl:cublas    | 26010  |
 | impl:egblas    | 9480   |
 | gpu:copy       | 6600   |
 | cpu:allocate   | 4882   |
 | gpu:allocate   | 54     |
 | gpu:cpu_to_gpu | 8      |
 | par:assign     | 2      |
 ---------------------------
```

**Comment by wichtounet** [comment]
Regarding the fact that MSE still takes a lot of time in the timers it's normal. CUDA Kernels are asynchronous. But for MSE, we have at least one synchronization to CPU to get the final result (CUDA copies implies synchronization by default). And a synchronization will wait for all the kernels in the queue to be executed. This means that the timers are completely skewed towards synchronization. Unfortunately, this means that the counters are not great in GPU mode but the nvprof results are more interesting. 

Regarding the counters, this is because the copy is made in egblas and I have no counters in egblas, only in ETL. These counters are pretty dumb, they do not count the actual copies from CUDA, they are just counters incremented from ETL. So if someplace, the counters missing, they will not be accurate.
